test/jug.ts

```diff
@@ -0,0 +1,529 @@
+    let timestamp = (await starknet.getBlock()).timestamp;
+
+    try {
+      // jug.file(ILK, "duty", 1);
+      await invoke(admin, jug, 'file_duty', { ilk: ILK, what: l2String('duty'), data: uint(1n) });
+    } catch (err: any) {
+      // vm.expectRevert("Jug/rho-not-updated");
+      expect(err.message).to.contain('Jug/rho-not-updated');
+    }
+    // vm.warp(block.timestamp - 1);
+    await starknet.devnet.setTime(timestamp - 1);
+    await starknet.devnet.createBlock();
+    // Invalid name
+    try {
+      // jug.file(ILK, "badWhat", 1);
+      await invoke(admin, jug, 'file_duty', {
+        ilk: ILK,
+        what: l2String('badWhat'),
+        data: uint(1n),
+      });
+    } catch (err: any) {
+      // vm.expectRevert("Jug/file-unrecognized-param");
+      expect(err.message).to.contain('Jug/rho-not-updated');
+    }
+    // Not authed
+    try {
+      // jug.deny(address(this));
+      await invoke(admin, jug, 'deny', { usr: _admin });
+      // jug.file(ILK, "duty", 1);
+      await invoke(admin, jug, 'file_duty', {
+        ilk: ILK,
+        what: l2String('duty'),
+        data: uint(1n),
+      });
+    } catch (err: any) {
+      // vm.expectRevert("Jug/not-authorized");
+      expect(err.message).to.contain('Jug/not-authorized');
+    }
+  });
+
+  // function testInit() public {
+  it('test init', async () => {
+    // assertEq(rho(ILK), 0);
+    // assertEq(duty(ILK), 0);
+    expect(await rho(ILK)).to.be.equal(0n);
+    expect(await duty(ILK)).to.deep.equal(uint(0n));
+    // vm.expectEmit(true, true, true, true);
+    // emit Init(ILK);
+    // jug.init(ILK);
+    await invoke(admin, jug, 'init', { ilk: ILK });
+    // assertEq(rho(ILK), block.timestamp);
+    // assertEq(duty(ILK), RAY);
+    let timestamp = (await starknet.getBlock()).timestamp;
+    expect(await rho(ILK)).to.be.equal(BigInt(timestamp));
+    expect(await duty(ILK)).to.deep.equal(uint(RAY));
+  });
+
+  // function testDripUpdatesRho() public {
+  it('test drip updates rho', async () => {
+    // jug.init(ILK);
+    await invoke(admin, jug, 'init', { ilk: ILK });
+    // jug.file(ILK, "duty", 10 ** 27);
+    await invoke(admin, jug, 'file_duty', {
+      ilk: ILK,
+      what: l2String('duty'),
+      data: uint(RAY),
+    });
+    // jug.drip(ILK);
+    await invoke(admin, jug, 'drip', { ilk: ILK });
+    // assertEq(rho(ILK), block.timestamp);
+    let timestamp = (await starknet.getBlock()).timestamp;
+    expect(await rho(ILK)).to.be.equal(BigInt(timestamp));
+    // vm.warp(block.timestamp + 1);
+    await starknet.devnet.increaseTime(1);
+    await starknet.devnet.createBlock();
+    // assertEq(rho(ILK), block.timestamp - 1);
+    timestamp = (await starknet.getBlock()).timestamp;
+    expect(await rho(ILK)).to.be.equal(BigInt(timestamp - 1));
+    // jug.drip(ILK);
+    await invoke(admin, jug, 'drip', { ilk: ILK });
+    // assertEq(rho(ILK), block.timestamp);
+    timestamp = (await starknet.getBlock()).timestamp;
```

Comment:
Is it latest block? Why in line 286 it is called explicitly? Helper blockTimestamp() would be usefull here.